### PR TITLE
Clean up raise exception logic

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* #55692 Refactor C->C to C->R twice
* **#55656 Clean up raise exception logic**
* #55237 Enable fast gradcheck for real inputs and outputs
* #56279 Disable flaky tests

### For release notes
What: 
 - All errors that are silenced by "raise_exception=False" are now GradcheckError (which inherits from RuntimeError).
 
Why:
 - Due to a refactor of gradcheck

Workaround:
 - If you catch for 'RuntimeError' with `except RuntimeError`, since GradcheckError inherits from RuntimeError, no changes are necessary. However if you explicitly check for the errors type via `type(error)`, you'll need to update your code to check for `GradcheckError` instead.

Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Differential Revision: [D27920809](https://our.internmc.facebook.com/intern/diff/D27920809)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -446,7 +438,7 @@ def backward_fn(grad_output):
 def _get_analytical_jacobian(inputs, outputs, input_idx, output_idx):
     # Computes the analytical jacobian in slow mode for a single input_idx, output_idx pair
     # without performing checks for dtype, shape, and reentrancy
-    jacobians, _ = check_analytical_jacobian_attributes(inputs, outputs[output_idx], float('inf'), 1.0, False, False)
+    jacobians = check_analytical_jacobian_attributes(inputs, outputs[output_idx], float('inf'), 1.0, False)
```
Comment (albanD) [inline torch/autograd/gradcheck.py:441]: You removed two arguments in the function signature: raise_exception, custom_vjp_fn=None
But only one here? 
I think we need to start passing these as kwargs!

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -1034,47 +1015,46 @@ def gradcheck(
...
-        return False
-
-    tupled_inputs = _as_tuple(inputs)
+    # This is just a wrapper that handles the raise_exception logic
+    args = locals()
+    args.pop("raise_exception")
+    if not raise_exception:
+        try:
+            return gradcheck_helper(**args)
+        except RuntimeError:
```
Comment (albanD) [inline torch/autograd/gradcheck.py:1024]: This simplification is really nice.
I think you can make it even better by defining a custom Error type here like `GradCheckError`. And only catch that error in this except. That way, you won't intercept any of the user errors

**Review by albanD — COMMENTED** [review]
Just did a quick review, it looks mostly good.

Let me know once all the CI is good for a final review.

File: torch/autograd/gradcheck.py (hunk 1 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:441]: Its a bit confusing here for sure! I just wanted to remove custom_vjp_fn until we had some tests that actually need it

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -1019,46 +1004,45 @@ def gradcheck(
...
     """
-    def fail_test(msg):
-        if raise_exception:
-            raise RuntimeError(msg)
-        return False
-
-    tupled_inputs = _as_tuple(inputs)
+    # This is just a wrapper that handles the raise_exception logic
+    args = locals()
+    args.pop("raise_exception")
```
Comment (albanD) [inline torch/autograd/gradcheck.py:1009]: Does that modify the actual `locals` dict inplace? Or is it a shallow copy that you modify?

**Review by albanD — COMMENTED** [review]
This is a great change! :)

I think we still want to mark this as BC-breaking though as we change the type of the error raised by gradcheck when it fails.
But that is really minimal.

File: torch/autograd/gradcheck.py (hunk 3 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:1009]: We are actually modifying the locals dict inplace. At first that seemed okay, but 
from https://nedbatchelder.com/blog/201211/tricky_locals.html it looks like there are cases where the local symbol table is only copied into locals() after every statement. Good catch!

File: torch/autograd/gradcheck.py [hunk 4]
```diff
@@ -499,7 +497,7 @@ def check_inputs(fail_test, tupled_inputs, check_sparse_nnz) -> bool:
             # "any overlap in memory" once we have a proper function to check it.
             if content.layout is not torch._mkldnn:  # type: ignore
                 if not all(st > 0 or sz <= 1 for st, sz in zip(content.stride(), content.size())):
-                    raise RuntimeError(
+                    raise ValueError(
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/autograd/gradcheck.py:500]: Oh right, this would be bc-breaking as well

**Review by albanD — APPROVED** [review]
LGTM!

**Pull request closed without merging**